### PR TITLE
MF-739 - Use User ID as a reference to an owner instead of an email

### LINKED ISSUE #739: Database design - add ID to the User entity and use it in Owner fields of other entities
Looking at Users and Things services i discovered that database design can be improved.
Mainflux uses `Emails` as primary key that according to database design patterns and methods is not a good practice due to slower performance ( because of string comparison),scalability and ....
One other thing is that repeating email address as `Channels` and `Things` owner is not size efficient and of course using it as primary key together with channel and things IDs just cause system slower on finding matches in queries.
I looking forward to make a Pull request for design a more performant database for storing persistent datas.
please also Take a look at [Cockroachdb](https://www.cockroachlabs.com/) as a better alternative to Postgresql than can do more with the same query language.

### PR DESCRIPTION
**Pull request by mteodor**
### What does this do?
Replace `email`  as owner reference in `things` and `channels` with user ID.
Using ID as a reference makes it easier to integrate with external source of user identity. 

Additionally I've created additional `register` endpoint for self-registering user, this should be separated from creating user by admin as it requires different request validation and different policy

### List any changes that modify/break current functionality

`owner` column in `things` and `channels` needs to be updated to corresponding user id

since things and users are separate database we can use `postgres_fdw` to make connections between database servers and updata properly
```
# connect to database things 
docker exec -it mainflux-things-db psql -U mainflux -d things
# connect to database things
\c things

# create users server 
create server users_server foreign data wrapper postgres_fdw options(host 'mainflux-users-db', dbname 'users', port '5432');

create user mapping for postgres server users_server options (user 'postgres', password 'xxxx');

import foreign schema public limit to(users) from server users_server into public;
```

now it is possible to cross reference tables from `users` and `things` database and update things and channels
```
docker exec -it mainflux-things-db psql -U mainflux -d things -c "create extension postgres_fdw";
docker exec -it mainflux-things-db psql -U mainflux -d things -c "create server users_server foreign data wrapper postgres_fdw options(host 'mainflux-users-db',dbname 'users', port '5432');"
docker exec -it mainflux-things-db psql -U mainflux -d things -c "create user mapping for mainflux server users_server options (user 'mainflux', password 'mainflux');"
docker exec -it mainflux-things-db psql -U mainflux -d things -c "import foreign schema public limit to(users) from server users_server into public;"
docker exec -it mainflux-things-db psql -U mainflux -d things -c "update things as t set owner=cast(u.id as varchar) from users u where u.email = t.owner;"
docker exec -it mainflux-things-db psql -U mainflux -d things -c "select * from things;"

```

### Notes

Resolves #739

### CODE REVIEW

File: pkg/sdk/go/groups.go [hunk 1]
```diff
@@ -126,7 +126,11 @@ func (sdk mfSDK) Unassign(token, groupID string, memberIDs ...string) error {
 	return nil
 }
 
+<<<<<<< HEAD
```
Comment (arvindh123) [inline pkg/sdk/go/groups.go]: Git merge lines in code , These lines have any purpose ?

File: pkg/sdk/go/users_test.go [hunk 2]
```diff
@@ -60,14 +67,14 @@ func TestCreateUser(t *testing.T) {
...
 
-	mockAuthzDB := map[string][]mocks.SubjectSet{}
-	mockAuthzDB[user.Email] = append(mockAuthzDB[user.Email], mocks.SubjectSet{Object: "authorities", Relation: "member"})
-	auth := mocks.NewAuthService(map[string]string{user.Email: user.Email}, mockAuthzDB)
-
-	tkn, _ := auth.Issue(context.Background(), &mainflux.IssueReq{Id: user.ID, Email: user.Email, Type: 0})
-	token := tkn.GetValue()
+	// mockAuthzDB := map[string][]mocks.SubjectSet{}
+	// mockAuthzDB[user.Email] = append(mockAuthzDB[user.Email], mocks.SubjectSet{Object: "authorities", Relation: "member"})
+	// auth := mocks.NewAuthService(map[string]users.User{userEmail: user}, mockAuthzDB)
```
Comment (manuio) [inline pkg/sdk/go/users_test.go]: Should we remove this commented code?

File: users/api/endpoint_test.go [hunk 3]
```diff
@@ -217,15 +229,15 @@ func TestUser(t *testing.T) {
...
-	auth := mocks.NewAuthService(map[string]string{user.Email: user.Email}, mockAuthzDB)
-	tkn, _ := auth.Issue(context.Background(), &mainflux.IssueReq{Id: user.ID, Email: user.Email, Type: 0})
-	token := tkn.GetValue()
+	auth := mocks.NewAuthService(map[string]users.User{}, mockAuthzDB)
 
-	userID, err := svc.Register(context.Background(), token, user)
+	userID, err := svc.Register(context.Background(), "", user)
 	require.Nil(t, err, fmt.Sprintf("register user got unexpected error: %s", err))
 
+	tkn, _ := auth.Issue(context.Background(), &mainflux.IssueReq{Id: user.ID, Email: user.Email, Type: 0})
```
Comment (manuio) [inline users/api/endpoint_test.go]: Check the error here please

**Review by manuio — CHANGES_REQUESTED** [review]

File: cli/users.go [hunk 4]
```diff
@@ -56,6 +56,32 @@ var cmdUsers = []cobra.Command{
 			logJSON(u)
 		},
 	},
+	{
+		Use:   "register",
+		Short: "register <username> <password> ",
+		Long:  `Registers new user`,
+		Run: func(cmd *cobra.Command, args []string) {
+			if len(args) < 2 || len(args) > 2 {
```
Comment (manuio) [inline cli/users.go]: Why not `if len(args) != 2` ?

**Review by manuio — CHANGES_REQUESTED** [review]

**Comment by drasko** [comment]
@mteodor this looks very nice. We will review as soon as we finish with these broker abstractions.

**Comment by mteodor** [comment]
> @mteodor this looks very nice. We will review as soon as we finish with these broker abstractions.

@drasko I hoped that this would be merged before Vidovdan so I could celebrate properly, but to no avail

**Comment by mteodor** [comment]
closed due to inactivity

**Pull request closed without merging**